preprocessing/__init__.py

```diff
@@ -81,6 +82,84 @@ def check_complement(g: Graph, h: Graph) -> (Graph, Graph):
         return g, h
 
 
+def find_components(g: Graph):
+    """
+    Breadth First Search Alg. Which also:
+    tests if the graph is connected,
+    computes the distance from s to the other edges
+    labels the vertices in the order they are visited
+    :param g: Graph
+    :return: (isConnected, {dict of components})
+    """
+    visited = set()
+    components = dict()
+    count = 1
+
+    while len(visited) < len(g.vertices):
+        for o in g.vertices:
+            if o not in visited:
+                v = o
+                break
+        queue = DoubleLinkedList()  # queue
+        queue.append(v)
+        visited.add(v)
+        components[count] = [v]
+
+        while len(queue) > 0:
+            w = queue.pop()  # BFS so FIFO
+            for n in w.neighbours:
+                # If w not visited
+                if n not in visited:
+                    visited.add(n)
+                    queue.append(n)
+                    components[count].append(n)
+        count += 1
+
+    is_connected = len(components) == 1
+    return is_connected, components
+
+
+def construct_graph_from_components(components: dict) -> [Graph]:
+    """
+    constructs a list of graphs from a dictionary of components
+
+        :param components: a dictionary constructed from an (unconnected) graph
+        :return: list of subgraphs
+        """
+    graphs = list()
+    for key in components.keys():
+        subgraph = Graph(False)
+        for vertex in components[key]:
+            if vertex not in subgraph.vertices:
+                subgraph.add_vertex(vertex)
+                for edge in vertex.incidence:
+                    if edge not in subgraph.edges:
+                        subgraph.add_edge(edge)
+        graphs.append(subgraph)
+    return graphs
+
+
+def compare_graph_components(g: [Graph], h: [Graph]) -> bool:
+    """
+    compares subgraphs in 2 graphs
+
+        :param g: a list of subgraphs
+        :param h: a list of subgraphs
+        :return: boolean: True if all preprocessing checks pass
+        """
+    if len(g) != len(h):
+        return False
+
+    for g_subgraph in g:
+        subgraph_equality = False
+        for h_subgraph in h:
+            if checks(g_subgraph, h_subgraph):
+                subgraph_equality = True
+        if not subgraph_equality:
+            return False
+    return True
+
```

Comment:
Kunnen we hier alvast paartjes maken van de subgraphs die overeenkomen met elkaar, zodat we die daarna slim kunnen verwerken? - AG